app/src/main/java/com/sunasterisk/food_01/data/source/remote/api/GetJsonFromUrl.kt

```diff
@@ -0,0 +1,31 @@
+package com.sunasterisk.food_01.data.source.remote.api
+
+import android.os.AsyncTask
+import com.sunasterisk.food_01.data.source.remote.OnFetchDataJsonListener
+import org.json.JSONObject
+
+class GetJsonFromUrl<T> constructor(private val listener: OnFetchDataJsonListener<T>,
+                                    private val keyEntity: String) : AsyncTask<String?, Void?, String?>() {
+
+    private var exception: Exception? = null
+
+    override fun doInBackground(vararg strings: String?): String? {
+        var data = ""
+        try {
+            val parseDataWithJson = ParseDataWithJson()
+            data = parseDataWithJson.getJsonFromUrl(strings[0]).toString()
+        } catch (e: Exception) {
+            exception = e
+        }
+        return data
+    }
+
+    override fun onPostExecute(data: String?) {
+        super.onPostExecute(data)
+        if (data != null && !data.isBlank()) {
+            val jsonObject = JSONObject(data)
+            @Suppress("UNCHECKED_CAST")
+            listener.onSuccess(ParseDataWithJson().parseJsonToData(jsonObject, keyEntity) as T)
+        } else listener.onError(exception)
+    }
+}
```

Comment:
```suggestion

    override fun doInBackground(vararg strings: String?): String? {
        try {
            return ParseDataWithJson().getJsonFromUrl(strings[0]).toString()
        } catch (e: Exception) {
         listener.onError(e)
        }
        return null
    }

    @Suppress("UNCHECKED_CAST")
    override fun onPostExecute(data: String?) {
        super.onPostExecute(data)
        if (data != null && !data.isBlank()) {
            listener.onSuccess(
            ParseDataWithJson().parseJsonToData(JSONObject(data), keyEntity) as T
            )
        }
    }
}
```
viết thế này được không ?

Reply:
theo e nghĩ thì hàm doInBackground là hàm thực thi nên câu lệnh này  /* listener.onError(e) */ nằm ở onPostExecute